daemon/algod/api/server/lib/middlewares/auth_test.go

```diff
@@ -0,0 +1,186 @@
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package middlewares
+
+import (
+	"errors"
+	"net/http"
+	"testing"
+
+	"github.com/labstack/echo/v4"
+	"github.com/stretchr/testify/require"
+)
+
+var errSuccess = errors.New("unexpected success")
+var invalidTokenError = echo.NewHTTPError(http.StatusUnauthorized, InvalidTokenMessage)
+var e = echo.New()
+var testAPIHeader = "API-Header-Whatever"
+
+// success is the "next" handler, it is only called when auth allows the request to continue
+func success(ctx echo.Context) error {
+	return errSuccess
+}
+
+func TestAuth(t *testing.T) {
+	tokens := []string{"token1", "token2"}
+
+	tests := []struct {
+		name           string
+		url            string
+		header         string
+		token          string
+		method         string
+		expectResponse error
+		finalPath      string
+	}{
+		{
+			"Valid token (1)",
+			"N/A",
+			testAPIHeader,
+			tokens[0],
+			"GET",
+			errSuccess,
+			"",
+		},
+		{
+			"Valid token (2)",
+			"N/A",
+			testAPIHeader,
+			tokens[1],
+			"GET",
+			errSuccess,
+			"",
+		},
+		{
+			"Valid token Bearer Format (1)",
+			"N/A",
+			"Authorization",
+			"Bearer " + tokens[0],
+			"GET",
+			errSuccess,
+			"",
+		},
+		{
+			"Valid token Bearer Format (2)",
+			"N/A",
+			"Authorization",
+			"Bearer " + tokens[1],
+			"GET",
+			errSuccess,
+			"",
+		},
+		{
+			"Invalid token",
+			"N/A",
+			testAPIHeader,
+			"invalid_token",
+			"GET",
+			invalidTokenError,
+			"",
+		},
+		{
+			"Invalid token Bearer Format",
+			"N/A",
+			"Authorization",
+			"Bearer invalid_token",
+			"GET",
+			invalidTokenError,
+			"",
+		},
+		{
+			"Missing token",
+			"N/A",
+			"",
+			"",
+			"GET",
+			invalidTokenError,
+			"",
+		},
+		{
+			"Invalid token + OPTIONS",
+			"N/A",
+			testAPIHeader,
+			"invalid_token",
+			"OPTIONS",
+			errSuccess,
+			"",
+		},
+		{
+			"Invalid bearer token + OPTIONS",
+			"N/A",
+			"Authorization",
+			"Bearer invalid_token",
+			"OPTIONS",
+			errSuccess,
+			"",
+		},
+		{
+			"Token in url (1)",
+			"http://my-node.com:80/urlAuth/" + tokens[0] + "/v2/status",
+			"",
+			tokens[0],
+			"GET",
+			errSuccess,
+			"/v2/status",
+		},
+		{
+			"Token in url (2)",
+			"http://my-node.com:80/urlAuth/" + tokens[1] + "/v2/status",
+			"",
+			tokens[1],
+			"GET",
+			errSuccess,
+			"/v2/status",
+		},
+		{
+			"Invalid token in url",
+			"http://my-node.com:80/urlAuth/invalid_token/v2/status",
+			"",
+			"invalid_token",
+			"GET",
+			invalidTokenError,
+			"",
+		},
+	}
+
+	authFn := MakeAuth(testAPIHeader, tokens)
+	handler := authFn(success)
+
+	for _, test := range tests {
```

Comment:
use the testing sub-test method T.Run ?
https://golang.org/pkg/testing/#T.Run

Reply:
Neat. Updated.